### PR TITLE
Add argon2 password strategy + spec

### PR DESCRIPTION
**Pull request by Copywright**
I wanted to add argon2 strategy for a project I'm using.

I'm up against a hiccup though, `Argon2::Password.create("password")` gives me: `NoMethodError: undefined method `create' for Argon2::Password:Class`.

Using [ruby-argon2](https://github.com/technion/ruby-argon2). 

I'm still digging in, would love another set of eyes.

EDIT: See the contributing guide, fixing the spec.

EDIT 2: Done!

### CODE REVIEW

**Comment by Copywright** [comment]
Not sure if this is due to my edit:

`>> BUNDLE_GEMFILE=/Users/herbertjoseph/Webdev/open_source/clearance/gemfiles/rails50.gemfile bundle exec rake
rake aborted!
NoMethodError: undefined method `last_comment' for #<Rake::Application:0x007f8172dd15f8>
/Users/herbertjoseph/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.5.0.beta1/lib/rspec/core/rake_task.rb:91:in `define'
/Users/herbertjoseph/.rvm/gems/ruby-2.3.0/gems/rspec-core-3.5.0.beta1/lib/rspec/core/rake_task.rb:71:in `initialize'
/Users/herbertjoseph/Webdev/open_source/clearance/Rakefile:14:in `new'
/Users/herbertjoseph/Webdev/open_source/clearance/Rakefile:14:in `<top (required)>'
/Users/herbertjoseph/.rvm/gems/ruby-2.3.0/bin/ruby_executable_hooks:15:in `eval'
/Users/herbertjoseph/.rvm/gems/ruby-2.3.0/bin/ruby_executable_hooks:15:in `<main>`

File: lib/clearance/password_strategies/argon2.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+module Clearance
+  module PasswordStrategies
+    module Argon2
+      require 'argon2'
+      require 'pry'
```
Comment (derekprior) [inline lib/clearance/password_strategies/argon2.rb]: You'll need to delete this here.

File: lib/clearance/password_strategies/argon2.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+module Clearance
+  module PasswordStrategies
+    module Argon2
+      require 'argon2'
```
Comment (derekprior) [inline lib/clearance/password_strategies/argon2.rb]: `require` should happen at the top of the file.

File: lib/clearance/password_strategies/argon2.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      # @api private
+      def encrypt(password)
+        t_cost, m_cost = costs
+        argon = ::Argon2::Password.new(t_cost: t_cost, m_cost: m_cost)
+        argon.create(password)
+      end
+
+      # @api private
+      def costs
+        test_environment? ? [1, 4] : [2, 16]
```
Comment (derekprior) [inline lib/clearance/password_strategies/argon2.rb]: What do you think about extracting constant for these arrays? 

```
MINIMAL_COST = [1, 4]
DEFAULT_COST = [2, 16]
```

Additionally, we do not use ternaries in this code base as we prefer the explicitness of branching conditionals:

```
if test_environment?
  MINIMAL_COST
else
  DEFAULT_COST
end
```

**Comment by derekprior** [comment]
Hello @Copywright. Thanks for preparing this. Argon2 has been on my list of things to investigate for Clearance 2.0, but I haven't had the time to do it. I was excited to see this come across.

I've given some feedback on the code, but ultimately I don't think I can accept this into clearance directly at the moment. There are a few reasons:

1. I don't want to add a dependency on `argon2` while most users of the library are likely to be using bcrypt.

2. I prefer, at least for now, to ship a single password strategy that we recommend. We currently have some other password strategies in the repo, but they are deprecated.

3. I'd actually like to change the implementation of PW strategies for Clearance 2. Right now, they get mixed into your `User` object directly but I really don't want to to do that anymore if we can help it.

I like what you've done here, though. What I'd suggest is extracting the work you've done here into a gem, `clearance-argon2`, that depends on Clearance 1.x. Then you'll be able to use argon2 and we can monitor how it's working for people as we decide how to move forward with strategies in Clearance 2. It's possible that we'll pull this in for 2.0 in some form, but I'm just not sure yet.

If you need assistance in how you would go about setting up the stand-alone gem, you can take a peek at [clearance-deprecated_password_strategies](https://github.com/thoughtbot/clearance-deprecated_password_strategies) to see how we pulled out the earlier strategies. If you need more help, let me know.

**Comment by Copywright** [comment]
@derekprior I understand, I'll do just that and make a gem abstraction

**Pull request closed without merging**